x-pack/plugins/security_solution/server/lib/detection_engine/signals/threat_mapping/get_signals_map_from_threat_index.ts

```diff
@@ -0,0 +1,126 @@
+  ThreatTermNamedQuery,
+  ThreatListItem,
+  SignalValuesMap,
+} from './types';
+import { getThreatList } from './get_threat_list';
+import { decodeThreatMatchNamedQuery } from './utils';
+
+import { MAX_NUMBER_OF_SIGNAL_MATCHES } from './enrich_signal_threat_matches';
+
+export type SignalsQueryMap = Map<string, ThreatMatchNamedQuery[]>;
+
+interface GetSignalsMatchesFromThreatIndexOptions {
+  threatSearchParams: Omit<GetThreatListOptions, 'searchAfter'>;
+  eventsCount: number;
+  signalValueMap?: SignalValuesMap;
+}
+
+/**
+ * fetches threats and creates signals map from results, that matches signal is with list of threat queries
+ */
+export const getSignalsQueryMapFromThreatIndex = async ({
+  threatSearchParams,
+  eventsCount,
+  signalValueMap,
+}: GetSignalsMatchesFromThreatIndexOptions): Promise<SignalsQueryMap> => {
+  let threatList: Awaited<ReturnType<typeof getThreatList>> | undefined;
+  const signalsQueryMap = new Map<string, ThreatMatchNamedQuery[]>();
+  // number of threat matches per signal is limited by MAX_NUMBER_OF_SIGNAL_MATCHES. Once it hits this number, threats stop to be processed for a signal
+  const maxThreatsReachedMap = new Map<string, boolean>();
+
+  const addSignalValueToMap = ({
+    signalId,
+    threatHit,
+    decodedQuery,
+  }: {
+    signalId: string;
+    threatHit: ThreatListItem;
+    decodedQuery: ThreatMatchNamedQuery | ThreatTermNamedQuery;
+  }) => {
+    const signalMatch = signalsQueryMap.get(signalId);
+    if (!signalMatch) {
+      signalsQueryMap.set(signalId, []);
+    }
+
+    const threatQuery = {
+      id: threatHit._id,
+      index: threatHit._index,
+      field: decodedQuery.field,
+      value: decodedQuery.value,
+      queryType: decodedQuery.queryType,
+    };
+
+    if (!signalMatch) {
+      signalsQueryMap.set(signalId, [threatQuery]);
+      return;
+    }
+
+    if (signalMatch.length === MAX_NUMBER_OF_SIGNAL_MATCHES) {
+      maxThreatsReachedMap.set(signalId, true);
+    } else if (signalMatch.length < MAX_NUMBER_OF_SIGNAL_MATCHES) {
+      signalMatch.push(threatQuery);
+    }
+  };
+
+  while (
+    maxThreatsReachedMap.size < eventsCount &&
+    (threatList ? threatList?.hits.hits.length > 0 : true)
+  ) {
+    threatList = await getThreatList({
+      ...threatSearchParams,
+      searchAfter: threatList?.hits.hits[threatList.hits.hits.length - 1].sort || undefined,
+    });
+
+    threatList.hits.hits.forEach((threatHit) => {
+      const matchedQueries = threatHit?.matched_queries || [];
+
+      matchedQueries.forEach((matchedQuery) => {
+        const decodedQuery = decodeThreatMatchNamedQuery(matchedQuery);
+        const signalId = decodedQuery.id;
+
+        if (decodedQuery.queryType === ThreatMatchQueryType.term) {
+          const threatValue = get(threatHit?._source, decodedQuery.value);
+          const values = Array.isArray(threatValue) ? threatValue : [threatValue];
+
+          values.forEach((value) => {
+            if (value && signalValueMap) {
```

Comment:
It appears that `signalValueMap` is required for this function to work correctly when the query against the `threatList` uses a `terms` query. However, the type system doesn't enforce that constraint so it would be easy to miss and accidentally use a `terms` query without providing the `signalValueMap`.

We should continue to work on ways to simplify the threat match rule logic so it's easier to maintain moving forward, and use the TS compiler to enforce requirements like this whenever possible.

Reply:
Good catch.
This part was adopted from the initial [IM terms PR ](https://github.com/elastic/kibana/pull/144511) and integrated into this one.
The issue I see here, it's difficult to determine whether query has any terms clauses within it and whether they related to IM execution path. Also it can be determined only in a runtime environment, when all checks for index fields performed and it's clear terms query can be generated.

So, instead, I introduced parameter `termsQueryAllowed`, that would require from developer to set it to `true` when terms query can be used. Once its set, it will require `signalValueMap` to be passed as argument